File: python-lib/adk_backend/routes/adoption.py
"""Adoption / engagement analytics — persistent spine (v1, pure DSS API).

The whole feature rests on a two-speed data model. This v1 route uses only the
*persistent* layer: project git history (author == DSS login, unbounded to
years), user `creationDate`, and the `list_users_activity()` snapshot. No audit
access, no host macro — the git logs it aggregates are already fetched and cached
by the projects catalog, so this is a cached read + a pure roll-up.

Two deeper layers live behind host macros, each with its own endpoint so
window-honesty stays structural rather than a disclaimer:
- /api/adoption/inventory — config-tree object inventory (full history of
  SURVIVING objects; deleted work is invisible — survivorship bias).
- /api/adoption/events — audit-log msgType event mix (whatever window the
  rotated audit files still cover).
"""
import logging
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from flask import Blueprint, g, jsonify

from adk_backend.caching import _cache_get
from adk_backend.clients import _adoption_git_aggregate, _sdk_fetch
from adk_backend.macros import _adoption_events_macro, _adoption_inventory_macro
from adk_backend.settings import _BACKEND_SETTINGS, _outreach_thresholds

bp = Blueprint('adoption', __name__)
_LOGGER = logging.getLogger(__name__)

_ADOPTION_CACHE_KEY = 'adoption'
_ADOPTION_INVENTORY_CACHE_KEY = 'adoption_inventory'
_ADOPTION_EVENTS_CACHE_KEY = 'adoption_events'
# Recent-activity pulse: ask for 3 days back; the tail-scan stops wherever the
# rotated files run out and reports what it actually covered.
_PULSE_WINDOW_HOURS = 72
_PULSE_TTL_SECONDS = 60


def _creation_month(created_ms: Any) -> Optional[str]:
    """'YYYY-MM' onboarding cohort bucket for a user creationDate (ms epoch)."""
    try:
        dt = datetime.fromtimestamp(int(created_ms) / 1000.0, tz=timezone.utc)
        return "%04d-%02d" % (dt.year, dt.month)
    except (TypeError, ValueError, OverflowError, OSError):
        return None


def _activity_dict(activity: Any) -> Dict[str, Any]:
    """`list_users_activity()` yields DSSUserActivity objects whose payload is the
    `.activity` dict; tolerate a raw dict too."""
    inner = getattr(activity, 'activity', None)
    if isinstance(inner, dict):
        return inner
    return activity if isinstance(activity, dict) else {}


def _licensing_summary(client: Any) -> Optional[Dict[str, Any]]:
    """Licensed seat limits + validity from get_licensing_status(), or None when
    the key can't read it. Shape verified live (akaos, DSS 14): limits live in
    limits.licensedProfiles[profile].licensedLimit (-1 = no limit); there are NO
    per-profile usage counts in the payload — usage joins from list_users."""
    try:
        lic = client.get_licensing_status() or {}
    except Exception:
        return None
    base = lic.get('base') or {}
    content = base.get('licenseContent') or {}
    profiles = []
    for name, row in ((lic.get('limits') or {}).get('licensedProfiles') or {}).items():
        if not isinstance(row, dict):
            continue
        profiles.append({'profile': name, 'licensedLimit': row.get('licensedLimit')})
    return {
        'valid': bool(base.get('valid')),
        'expired': bool(base.get('expired')),
        'expiresOnMs': base.get('expiresOn'),
        'licenseKind': content.get('licenseKind'),
        'communityEdition': bool(base.get('community')),
        'profiles': profiles,
    }


def _adoption_data(client: Any) -> Dict[str, Any]:
    """Assemble the persistent adoption payload: git spine + active/total (matched
    to the Inactive-Projects threshold) + onboarding cohorts + per-user recency."""
    agg = _adoption_git_aggregate(client)
    project_rows: List[Dict[str, Any]] = agg.get('projectRows', [])

    # Active vs total projects (#5) — same rule as /api/tools/inactive-projects so
    # the headline KPI matches that tool exactly (last activity within threshold).
    threshold_days = _outreach_thresholds.get('inactive_project_days', 180)
    now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
    active_count = 0
    for row in project_rows:
        last_ms = row.get('lastCommitMs')
        is_active = False
        if last_ms is not None:
            days = (now_ms - int(last_ms)) / (1000 * 60 * 60 * 24)
            is_active = days < threshold_days
        row['active'] = is_active
        if is_active:
            active_count += 1

    users = _sdk_fetch(
        'list_users',
        _BACKEND_SETTINGS['cache_ttl_users'],
        lambda: client.list_users(),
    ) or []
    activities = _sdk_fetch(
        'list_users_activity',
        _BACKEND_SETTINGS['cache_ttl_users'],
        lambda: client.list_users_activity(),
    ) or []

    # Onboarding cohorts (#) — user creationDate bucketed by month.
    cohort_counts: Dict[str, int] = {}
    creation_by_login: Dict[str, Any] = {}
    display_by_login: Dict[str, str] = {}
    profile_by_login: Dict[str, str] = {}
    profile_counts: Dict[str, int] = {}
    groups_by_login: Dict[str, List[str]] = {}
    for u in users:
        login = u.get('login') or ''
        if not login:
            continue
        display_by_login[login] = u.get('displayName') or login
        profile = u.get('userProfile') or ''
        profile_by_login[login] = profile
        if profile:
            profile_counts[profile] = profile_counts.get(profile, 0) + 1
        groups_by_login[login] = [g for g in (u.get('groups') or []) if g]
        created = u.get('creationDate')
        creation_by_login[login] = created
        month = _creation_month(created)
        if month:
            cohort_counts[month] = cohort_counts.get(month, 0) + 1
    cohorts = [{'month': m, 'newUsers': cohort_counts[m]} for m in sorted(cohort_counts.keys())]

    # Most active groups — roll each builder's git activity up to their DSS
    # groups (`user.groups` is a plain list of names). A builder in N groups
    # counts toward all N (commit shares can overlap, like the native DSS
    # group model itself). Builders whose account is gone have no groups row —
    # their work still counts in totals/trend, just not here.
    builder_stats: List[Dict[str, Any]] = agg.get('builderStats', [])
    builder_projects: Dict[str, Any] = agg.get('builderProjects', {})
    builder_monthly: Dict[str, Dict[str, int]] = agg.get('builderMonthlyCommits', {})
    stats_by_login = {b['login']: b for b in builder_stats}
    group_rows: Dict[str, Dict[str, Any]] = {}
    for login, group_names in groups_by_login.items():
        stat = stats_by_login.get(login)
        for name in group_names:
            row = group_rows.setdefault(name, {
                'name': name, 'memberCount': 0, 'builderCount': 0, 'commits': 0,
                'projects': set(), 'lastCommitMs': None, 'monthlyCommits': {},
            })
            row['memberCount'] += 1
            if stat is None:
                continue
            row['builderCount'] += 1
            row['commits'] += stat['commits']
            row['projects'].update(builder_projects.get(login, ()))
            monthly = row['monthlyCommits']
            for month, n in builder_monthly.get(login, {}).items():
                monthly[month] = monthly.get(month, 0) + n
            last = stat.get('lastCommitMs')
            if last is not None and (row['lastCommitMs'] is None or last > row['lastCommitMs']):
                row['lastCommitMs'] = last
    groups = []
    for row in group_rows.values():
        row['projectCount'] = len(row.pop('projects'))
        groups.append(row)
    groups.sort(key=lambda r: (-r['commits'], -r['memberCount'], r['name']))

    # Top builders — leaderboard with display names joined in (a departed
    # builder keeps their login as the display name).
    top_builders = [
        {**b, 'displayName': display_by_login.get(b['login'], b['login'])}
        for b in builder_stats
    ]

    # Login recency reframed (#1) — persistent snapshot presented as "last active",
    # never a fabricated login count (login-success events undercount 10-100x).
    builder_recency: List[Dict[str, Any]] = []
    for a in activities:
        act = _activity_dict(a)
        login = act.get('login') or ''
        if not login:
            continue
        builder_recency.append({
            'login': login,
            'displayName': display_by_login.get(login, login),
            'lastSuccessfulLogin': act.get('lastSuccessfulLogin'),
            'lastSessionActivity': act.get('lastSessionActivity'),
            'creationDate': creation_by_login.get(login),
            'userProfile': profile_by_login.get(login) or None,
        })

    totals = dict(agg.get('totals', {}))
    totals['activeProjectCount'] = active_count
    totals['inactiveThresholdDays'] = threshold_days

    return {
        'ok': True,
        'generatedAtMs': now_ms,
        'totals': totals,
        'monthlyTrend': agg.get('monthlyTrend', []),
        'projectRows': project_rows,
        'cohorts': cohorts,
        'repeatBuilders': agg.get('repeatBuilders', {'total': 0, 'single': 0, 'repeat': 0}),
        # login → 'YYYY-MM' → commits, for the new/returning/lapsed lifecycle
        # view. Already aggregated by _adoption_git_aggregate for the group
        # roll-up — exposing it costs nothing extra.
        'builderMonthly': agg.get('builderMonthlyCommits', {}),
        'builderRecency': builder_recency,
        'groups': groups,
        'builderStats': top_builders,
        # Licensed seat limits (None when the key can't read licensing) + the
        # actual per-profile seat usage from the same list_users snapshot.
        'licensing': _licensing_summary(client),
        'profileCounts': profile_counts,
    }


@bp.route('/api/adoption')
def api_adoption():
    client = g.client

    def loader():
        return _adoption_data(client)

    ttl = int(_BACKEND_SETTINGS.get('cache_ttl_projects', 600))
    data = _cache_get(_ADOPTION_CACHE_KEY, ttl, loader)
    return jsonify(data)


@bp.route('/api/adoption/inventory')
def api_adoption_inventory():
    """Config-tree object inventory (macro): full history of surviving objects.

    Cache entries are scoped per active host: _cache_get routes every key
    through caching._cache_key, which prefixes the request's X-DSS-Host-Id
    (set on _THREAD_LOCAL by @before_request) — a remote host's inventory is
    never served for the local one, matching /api/adoption and /api/cru.
    """
    client = g.client

    def loader():
        return _adoption_inventory_macro(client)

    ttl = int(_BACKEND_SETTINGS.get('cache_ttl_projects', 600))
    data = _cache_get(_ADOPTION_INVENTORY_CACHE_KEY, ttl, loader)
    return jsonify(data)


@bp.route('/api/adoption/events')
def api_adoption_events():
    """Recent-activity pulse (macro, mode=recent): a cheap reverse tail-scan of
    the newest audit files, covering the last ~72h or however far the rotated
    files actually reach — the payload reports the MEASURED window, never the
    requested one (20 rotations can cover under 24h on a busy instance).

    Host-scoped cache: see api_adoption_inventory — _cache_key prefixes the
    active host id, so entries never cross hosts. Short TTL: this is the one
    fast-moving card on the page.
    """
    client = g.client

    def loader():
        return _adoption_events_macro(client, mode='recent', window_hours=_PULSE_WINDOW_HOURS)

    data = _cache_get(_ADOPTION_EVENTS_CACHE_KEY, _PULSE_TTL_SECONDS, loader)
    return jsonify(data)

File: resource/frontend/src/types/adoption.ts
// ── Adoption / engagement analytics (persistent spine, v1) ────────────────────
// Shape mirrors python-lib/adk_backend/routes/adoption.py JSON output exactly.
// Every field here is PERSISTENT (spans each project's whole git history / the
// user snapshot), never a short audit window — the UI frames it as such.

/** One month of the persistent adoption curve. */
export interface AdoptionMonthPoint {
  month: string; // 'YYYY-MM'
  activeBuilders: number; // distinct human git authors active that month
  commits: number; // human commit volume that month
}

/** Per-project people & activity, from distinct git-commit authors. */
export interface AdoptionProjectRow {
  projectKey: string;
  name: string;
  owner: string;
  authors: string[]; // distinct human logins (== DSS login)
  authorCount: number; // people per project (#6)
  commits: number; // human commits
  activeMonths: number; // distinct months with human activity
  firstCommitMs: number | null;
  lastCommitMs: number | null; // newest non-migration commit (matches Inactive-Projects)
  active: boolean; // last activity within the inactive-project threshold
  truncated: boolean; // history deeper than the paginated fetch cap — counts are floors
}

/** Onboarding cohort: users created in a given month. */
export interface AdoptionCohort {
  month: string; // 'YYYY-MM'
  newUsers: number;
}

/** Returning-builder split: authors active in exactly one vs. multiple months. */
export interface AdoptionRepeatBuilders {
  total: number;
  single: number; // active in exactly 1 distinct month
  repeat: number; // active in >= 2 distinct months
}

/** DSS group activity roll-up — a builder in N groups counts toward all N. */
export interface AdoptionGroupRow {
  name: string;
  memberCount: number; // users carrying this group
  builderCount: number; // members with git activity
  commits: number; // sum of member builders' commits (shares can overlap)
  projectCount: number; // distinct projects touched by member builders
  lastCommitMs: number | null;
  /** 'YYYY-MM' → member commits that month (summed over member builders). */
  monthlyCommits: Record<string, number>;
}

/** Per-builder leaderboard row (sorted by commits desc server-side). */
export interface AdoptionBuilderRow {
  login: string;
  displayName: string;
  commits: number;
  projectCount: number;
  activeMonths: number;
  firstCommitMs: number | null;
  lastCommitMs: number | null;
}

/** Per-user recency from list_users_activity() — "last active", never a count. */
export interface AdoptionBuilderRecency {
  login: string;
  displayName: string;
  lastSuccessfulLogin: number | null;
  lastSessionActivity: number | null;
  creationDate: number | null;
  userProfile?: string | null; // seat type, for the seat-type × creators card
}

export interface AdoptionTotals {
  projectCount: number;
  activeProjectCount: number; // active vs total projects (#5)
  builderCount: number; // distinct human authors across all projects
  automationCount: number; // distinct api:/no:auth identities (excluded from people)
  commitCount: number;
  firstCommitMs: number | null;
  lastCommitMs: number | null;
  avgPeoplePerProject: number;
  inactiveThresholdDays: number;
  truncatedProjectCount: number; // projects whose history exceeded the fetch cap
}

/** One licensed profile's seat cap — licensedLimit ≤ 0 means "no limit". */
export interface AdoptionLicenseProfile {
  profile: string;
  licensedLimit: number | null;
}

/** Licensed seat limits from get_licensing_status() (verified live: no usage
 * counts in that payload — usage comes from profileCounts below). Null when
 * the API key can't read licensing. */
export interface AdoptionLicensing {
  valid: boolean;
  expired: boolean;
  expiresOnMs: number | null;
  licenseKind: string | null;
  communityEdition: boolean;
  profiles: AdoptionLicenseProfile[];
}

export interface AdoptionData {
  ok?: boolean;
  error?: string;
  generatedAtMs?: number;
  totals?: AdoptionTotals;
  monthlyTrend?: AdoptionMonthPoint[];
  projectRows?: AdoptionProjectRow[];
  cohorts?: AdoptionCohort[];
  repeatBuilders?: AdoptionRepeatBuilders;
  /** login → 'YYYY-MM' → commits — feeds the new/returning/lapsed lifecycle. */
  builderMonthly?: Record<string, Record<string, number>>;
  builderRecency?: AdoptionBuilderRecency[];
  groups?: AdoptionGroupRow[];
  builderStats?: AdoptionBuilderRow[];
  licensing?: AdoptionLicensing | null;
  /** userProfile → seat count, from the same list_users snapshot. */
  profileCounts?: Record<string, number>;
}

// ── Config-tree object inventory (macro layer) ────────────────────────────────
// Shape mirrors python-runnables/adoption-inventory/runnable.py exactly, which
// is itself a port of the diag-parser inventory accumulator. Covers the
// instance's full multi-year history — but only for objects that still exist
// (survivorship bias; deleted work is invisible).

export type ObjectFamily =
  | 'dataset'
  | 'recipe-visual'
  | 'recipe-python'
  | 'recipe-sql'
  | 'recipe-r'
  | 'recipe-ml'
  | 'recipe-plugin'
  | 'recipe-other'
  | 'notebook'
  | 'sql-notebook'
  | 'webapp'
  | 'dashboard'
  | 'insight'
  | 'saved-model'
  | 'scenario'
  | 'prompt-studio'
  | 'wiki'
  | 'zone'
  | 'analysis'
  | 'mes'
  | 'other';

/** Save-count histogram from versionTag.versionNumber (0 = never re-saved
 * after creation). Buckets are inclusive: v1 ≤ 1 save, v2to5 = 2–5 saves, … */
export interface InventoryEditBuckets {
  v1: number;
  v2to5: number;
  v6to20: number;
  v21plus: number;
}

export interface InventoryFamilyStats {
  count: number;
  /** Objects carrying a usable creationTag (login + timestamp). Coverage is
   * surfaced as "tagged N of M" — older/programmatic objects may lack tags. */
  tagged: number;
  /** Raw object subtype counts (recipe `type`, webapp BOKEH/DASH, …). */
  subtypes: Record<string, number>;
  /** Sum of versionTag.versionNumber across tagged objects (save volume). */
  versionSum: number;
  editBuckets: InventoryEditBuckets;
}

/** One 'YYYY-MM' creation bucket. `creators` maps login → objects created that
 * month; views derive the distinct-creator count. */
export interface InventoryCreationMonth {
  month: string;
  total: number;
  byFamily: Partial<Record<ObjectFamily, number>>;
  creators: Record<string, number>;
}

export interface InventoryCreatorStats {
  /** Objects whose creationTag.lastModifiedBy is this login. */
  created: number;
  byFamily: Partial<Record<ObjectFamily, number>>;
  firstCreatedMs: number | null;
  lastCreatedMs: number | null;
  /** Latest creation or last-save timestamp attributed to this login. */
  lastEditMs: number | null;
  /** Objects this login last saved (versionTag) but did not create. */
  editedNotCreated: number;
  /** Objects whose most recent save (versionTag) was by this login. */
  saves: number;
}

export interface InventoryProjectStats {
  objectCount: number;
  byFamily: Partial<Record<ObjectFamily, number>>;
  /** login → objects created in this project. */
  creators: Record<string, number>;
  /** Objects whose last editor differs from their creator (both tagged). */
  handoffCount: number;
  /** Tagged objects never re-saved after creation (versionNumber ≤ 0). */
  savedOnce: number;
  versionSum: number;
  lastHumanEditMs: number | null;
  lastEditor: string | null;
  /** 'YYYY-MM' → objects whose last edit falls in that month. Staleness is
   * NOT pre-collapsed — views bucket this against `lastEditMs` at render
   * time so thresholds stay tunable without re-scanning. */
  lastEditMonthCounts: Record<string, number>;
}

export interface ObjectInventory {
  families: Partial<Record<ObjectFamily, InventoryFamilyStats>>;
  creationMonths: InventoryCreationMonth[];
  creators: Record<string, InventoryCreatorStats>;
  projects: Record<string, InventoryProjectStats>;
  /** Earliest surviving creationTag timestamp (TTFB cohort floor). */
  firstCreationMs: number | null;
  /** Latest edit anywhere — the reference "now" for staleness collapse. */
  lastEditMs: number | null;
  /** Objects ingested (config JSONs + meta-only paths). */
  scanned: number;
  /** Objects with a usable creationTag. */
  taggedObjects: number;
  errors: number;
  /** Always true from the macro (single blocking pass). */
  complete: boolean;
}

/** /api/adoption/inventory payload: the inventory plus the macro envelope. */
export interface AdoptionInventoryData extends Partial<ObjectInventory> {
  ok?: boolean;
  error?: string;
  generatedAtMs?: number;
  projectCount?: number;
}

// ── Audit-tail recent-activity pulse (macro layer, mode=recent) ───────────────
// Shape mirrors python-runnables/adoption-events/runnable.py `_parse_recent`.
// Window-honesty: firstEventMs/lastEventMs are the MEASURED span — rotated
// audit files often cover far less than the requested window.

export interface AdoptionPulseHour {
  hourMs: number; // epoch ms, floored to the hour
  events: number; // human events that hour
  humans: number; // distinct humans that hour
}

export interface AdoptionPulseData {
  ok?: boolean;
  error?: string;
  mode?: string;
  generatedAtMs?: number;
  windowHours?: number; // requested window
  firstEventMs?: number | null; // measured span start
  lastEventMs?: number | null;
  coverageHours?: number | null;
  hours?: AdoptionPulseHour[];
  /** build / run / explore / consume / other totals (classified server-side). */
  buckets?: Record<string, number>;
  /** Run-bucket msgType counts (jobs, scenarios, macros…), capped server-side. */
  runTypes?: Record<string, number>;
  topHumans?: Array<{ login: string; events: number }>;
  humansActive?: number;
  /** True when the rotation set ran out before reaching the requested window. */
  exhaustedFiles?: boolean;
  filesRead?: number;
  linesScanned?: number;
}
